### PR TITLE
VACMS-11516 vba show services description facility view

### LINKED ISSUE #11516: CMS: Show Services name and description VBA Facility node:view
## User Story or Problem Statement

As an editor, I want to see the name and description of the services on the VBA Facility node:view so that I can see what services are currently listed and how they are described.

## Acceptance Criteria

- [ ] Under SERVICES: the name of the service <Facility name> - <Service name> is displayed, followed by the description from the "VBA description" in the "VA Services" term.
  - [ ] The term description follows the pattern of other national content, displayed in a greyed out box with a tooltip.
Screen Shot 2022-11-21 at 2 41 28 PM (2)

- [ ] If the "VBA service description" field has content for a facility service, that content displays after the national content.

Screen Shot 2022-11-08 at 2 33 48 PM

## Implementation steps
Create a module for VBA.
Use the code from `appendHealthServiceTermDescriptionToVetCenter` in `va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php` as a model.

## Design principles

Veteran-centered
- [x] `Single source of truth`: Increase reliability and consistency of content on VA.gov by providing a single source of truth.
- [ ] `Accessible, plain language`: Provide guardrails and guidelines to ensure content quality.
- [ ] `Purposely structured content`: Ensure Content API can deliver content whose meaning matches its structure.
- [ ] `Content lifecycle governance`: Produce tools, processes and policies to maintain content quality throughout its lifecycle.

Editor-centered
- [ ] `Purpose-driven`: Create an opportunity to involve the editor community in VA’s mission and content strategy goals.
- [ ] `Efficient`: Remove distractions and create clear, straightforward paths to get the job done.
- [ ] `Approachable`: Offer friendly guidance over authoritative instruction.
- [ ] `Consistent`: Reduce user’s mental load by allowing them to fall back on pattern recognition to complete tasks.
- [ ] `Empowering`: Provide clear information to help editors make decisions about their work.

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Program`
- [ ] `Platform CMS Team`
- [ ] `Sitewide Crew`
- [ ] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public Websites`
- [x] `⭐️ Facilities`
- [ ] `⭐️ User support`

### PR DESCRIPTION
**Pull request by omahane**
## Description

Closes #11516.

## Testing done

Manually.

## Screenshots
Screen Shot 2022-11-29 at 8 42 50 PM

## QA steps

As user with an admin role
1. Visit the [Pre-Discharge Site at Wasilla, AK](https://pr11696-tnnufvgx2phkyoicb0eltr92vbvmzhm3.ci.cms.va.gov/node/42633)
   - [x] Validate that the local service descriptions are showing for VBA 1
   - [x] Validate that no local service description is showing for VBA 2
   - [x] Hover over the tooltip to see that that "Why can’t I edit this? This nationally-managed service description will display before the facility-specific service description." shows
   - [x] Validate that the VBA 3 local content is showing but the national content is missing.
2. Then edit the page
   - [x] Validate that the "Pre-Discharge Site at Wasilla, AK - VBA 1" service has a facility-specific service description.
   - [x] Validate that the "Pre-Discharge Site at Wasilla, AK - VBA 2" service has no facility-specific service descrition.
   - [x] Click "Add new service"
     - [ ] For Facility, choose "Pre-Discharge Site at Wasilla, AK"
     - [ ] For Section, choose
     - [ ] For Service Name, select "VBA 4".
     - [ ] For VBA service description, add a facility specific description of your choosing.
     - [ ] Click "Create Service"
   - [x]  Save the page
   - [x] Validate that the VBA 4 service shows with the national description.
   - [x] Validate that the VBA 4 service has a facility-specific description.

Screen Shot 2022-11-29 at 10 09 28 PM

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [x] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Platform CMS Team`
- [ ] `Sitewide program`
- [ ] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public websites`
- [x] `⭐️ Facilities`
- [ ] `⭐️ User support`

### CMS user-facing announcement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

File: docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  public function appendServiceTermDescriptionToVbaFacility(EntityViewAlterEvent $event):void {
+    if ($event->getDisplay()->getTargetBundle() === 'vba_facility') {
+      $build = &$event->getBuild();
+      $services = $build['field_vba_services'] ?? [];
+      foreach ($services as $key => $service) {
+        if (is_numeric($key) && !empty($service['#options'])) {
+          $service_node = $service['#options']['entity'];
+          // Get the content from the taxonomy term description field.
+          $referenced_term_id = $service_node->get('field_service_name_and_descripti')->getValue()['0']['target_id'];
+          $entity = $this->entityTypeManager->getStorage('taxonomy_term')->load($referenced_term_id);
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php:103]: I think this risky grab of assuming that getValue() is returning an array that will have a target_id  and then jumping to loading that can be avoided like this.
```suggestion
          $entities = $service_node->get('field_service_name_and_descripti')->referencedEntities();
          $entity = reset($entities);
```
Of course $entity could be FALSE if that field is empty, so youll need to bail out if if it is FALSE.

**Review by swirtSJW — CHANGES_REQUESTED** [review]
Some suggestions for making this less risky.

File: docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        if (is_numeric($key) && !empty($service['#options']['entity'])) {
+          $service_node = $service['#options']['entity'];
+          // Get the content from the taxonomy term description field.
+          $entities = $service_node->get('field_service_name_and_descripti')->referencedEntities();
+          // $referenced_term_id = $service_node->get('field_service_name_and_descripti')->getValue()['0']['target_id'];
+          if (!empty($entities)) {
+            // Load the term as an entity.
+            $entity = reset($entities);
+            // Create the view instance.
+            $view_builder = $this->entityTypeManager->getViewBuilder('taxonomy_term');
```
Comment (omahane) [inline docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php:108]: @swirtSJW Do I also need to check $entity before using it?

I see you did something like this here:
```
/**
 * Implements hook_simplesamlphp_auth_existing_user().
 *
 * @see hook_simplesamlphp_auth_existing_user()
 */
function va_gov_login_simplesamlphp_auth_existing_user($attributes) {
  $saml_mail = $attributes['adEmail'];
  $existing_users = \Drupal::service('entity_type.manager')->getStorage('user')->loadByProperties(['mail' => $saml_mail]);
  if ($existing_users) {
    $existing_user = is_array($existing_users) ? reset($existing_users) : FALSE;
    if ($existing_user) {
      return $existing_user;
    }
  }
  return FALSE;
}
```
docroot/modules/custom/va_gov_login/va_gov_login.module

Comment (omahane) [reply docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php]: I've added that check.

File: docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php (hunk 1 above)
Comment (omahane) [reply docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php]: Changed and reworded.

File: docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+   *
+   * @param \Drupal\core_event_dispatcher\Event\Entity\EntityViewAlterEvent $event
+   *   The entity view alter service.
+   */
+  public function appendServiceTermDescriptionToVbaFacility(EntityViewAlterEvent $event):void {
+    if ($event->getDisplay()->getTargetBundle() === 'vba_facility') {
+      $build = &$event->getBuild();
+      $services = $build['field_vba_services'] ?? [];
+      foreach ($services as $key => $service) {
+        if (is_numeric($key) && !empty($service['#options'])) {
```
[the remark this answers is not in the document]
Comment (omahane) [reply docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php]: Changed.

**Comment by omahane** [comment]
@swirtSJW This is ready for another going-over.

**Comment by omahane** [comment]
@laflannery @BlakeOrgan I have this ready for review. Can you let me know what you think?
https://pr11696-tnnufvgx2phkyoicb0eltr92vbvmzhm3.ci.cms.va.gov/anchorage-va-regional-benefit-office-at-anchorage-va-medical-center/locations/pre-discharge-site-at-wasilla-ak

**Comment by laflannery** [comment]
3 things:

1. On the view screen, could we remove the italics from the notice when there is not a national description? Italics is not recommended in the VA content style guide - it can be difficult to read in some cases.
   1. image
3. I am confused by VBA 3 on the edit screen - the dropdown says VBA 1, the description says VBA 3 and the title has nothing:
   1. image
4. As someone who can edit the national service descriptions, there is an overlapping issue with the edit icon and the tooltips:
   1. I'm not sure how much of an issue this is, how many people this would impact, but I at least wanted to bring it up.  I also have a larger, future goal to overhaul the tooltips in general FYI. 
   2. image

**Comment by omahane** [comment]
> 1. On the view screen, could we remove the italics from the notice when there is not a national description? Italics is not recommended in the VA content style guide - it can be difficult to read in some cases.

Yes, I'll change that to bold only.

>    1. image
> 2. I am confused by VBA 3 on the edit screen - the dropdown says VBA 1, the description says VBA 3 and the title has nothing:

This confusion has to do with the fact that I created the VBA Service with a term that existed (VBA 3). Then, I deleted the term, mostly to show what happens when we delete a term, showing the "Service not found" notice on node:view. This would have to be the cue to change the service choice on the edit screen.

Similarly, on the node:edit screen, the fact that the service has no service name after the facility name is an indication that something is amiss. 

The VBA 1 you're seeing is the form defaulting to the first item on the list for the purposes of selection. It hasn't actually been set to VBA 1 (yet) when you're seeing it in the form that way.

Does that make sense?

Granted, this is an edge case (deleted taxonomy term), but we wanted to account for it somehow.

**Comment by omahane** [comment]
@laflannery @BlakeOrgan I edited this locally. I haven't pushed my changes, as that will reset the preview environment, including wiping out my content changes (like the terms and VBA services). 

Do these changes look sufficient?
Screen Shot 2022-11-30 at 8 37 34 AM (2)

**Review by laflannery — DISMISSED** [review]
Looks good, thanks!

File: docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
+<?php
+
+namespace Drupal\va_gov_vba_facility\EventSubscriber;
+
+use Drupal\Component\Render\FormattableMarkup;
+use Drupal\core_event_dispatcher\EntityHookEvents;
+use Drupal\Core\Entity\EntityTypeManager;
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php:7]:
```suggestion
use Drupal\Core\Entity\EntityTypeManagerInterface;
```
If there is an interface available, that is the better choicce to use.

File: docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+use Drupal\Core\Render\RendererInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\Core\StringTranslation\TranslationInterface;
+use Drupal\va_gov_user\Service\UserPermsService;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Drupal\core_event_dispatcher\Event\Entity\EntityViewAlterEvent;
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php:13]: I'm going to push up an alpha sort of these use statements.
A useful VS code plugin if you don't have it
image

File: docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+   * UserPerms Service.
+   *
+   * @var \Drupal\va_gov_user\Service\UserPermsService
+   */
+  protected $userPermsService;
+
+  /**
+   * The entity manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManager
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php:32]: reference the interface to match the use.

File: docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+   */
+  public function appendServiceTermDescriptionToVbaFacility(EntityViewAlterEvent $event):void {
+    if ($event->getDisplay()->getTargetBundle() === 'vba_facility') {
+      $build = &$event->getBuild();
+      $services = $build['field_vba_services'] ?? [];
+      foreach ($services as $key => $service) {
+        $description = new FormattableMarkup('', []);
+        if (is_numeric($key) && !empty($service['#options']['entity'])) {
+          $service_node = $service['#options']['entity'];
+          $referenced_term_array = $service_node->get('field_service_name_and_descripti')->referencedEntities();
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_vba_facility/src/EventSubscriber/VbaFacilitySubscriber.php:102]: Going a little hungarian notation here... which I used to love, but a more modern drupallly pattern is to name plurals if unknown quantity and singular when you have one.
```suggestion
          $referenced_terms = $service_node->get('field_service_name_and_descripti')->referencedEntities();
```

**Review by swirtSJW — COMMENTED** [review]
This all looks good and was testing out.  However I am seeing some acccess denied errors on save.
image



Path: /node/51488. Drupal\Core\Http\Exception\CacheableAccessDeniedHttpException: The 'view any unpublished content' permission is required. in Drupal\Core\Routing\AccessAwareRouter->checkAccess() (line 118 of /var/lib/tugboat/docroot/core/lib/Drupal/Core/Routing/AccessAwareRouter.php).
--

```
Path: /node/51488. Drupal\Core\Http\Exception\CacheableAccessDeniedHttpException: The 'view any unpublished content' permission is required. in Drupal\Core\Routing\AccessAwareRouter->checkAccess() (line 118 of /var/lib/tugboat/docroot/core/lib/Drupal/Core/Routing/AccessAwareRouter.php).
```  

I believe this is due to not exempting from an access check somewhere along the line.  I am looking into it right now.


Path: /node/51488. Drupal\Core\Http\Exception\CacheableAccessDeniedHttpException: The 'view any unpublished content' permission is required. in Drupal\Core\Routing\AccessAwareRouter->checkAccess() (line 118 of /var/lib/tugboat/docroot/core/lib/Drupal/Core/Routing/AccessAwareRouter.php).
--


<br class="Apple-interchange-newline">

Path: /node/51488. Drupal\Core\Http\Exception\CacheableAccessDeniedHttpException: The 'view any unpublished content' permission is required. in Drupal\Core\Routing\AccessAwareRouter->checkAccess() (line 118 of /var/lib/tugboat/docroot/core/lib/Drupal/Core/Routing/AccessAwareRouter.php).
--


<br class="Apple-interchange-newline">

**Comment by swirtSJW** [comment]
Turns out the access denied error is not coming from this PR.  I just saw it on staging as well
image

it does not need to be fixed for this PR to pass.

**Review by swirtSJW — APPROVED** [review]
@omahane Nice work on this.  I made a couple minor commits mainly hoping to get this accross the line tonight, but then I remembered you has a style change related to the Tool tip location conflicting with the edit pencil icon.

push your changes in the morning and we can call this done.

**Review by swirtSJW — APPROVED** [review]
Approved and ready for merge, just waiting until after demo.

**Pull request merged**